File: packages/core/skills/frontmatter.test.ts
import { describe, it, expect } from "vitest";
import { parseFrontmatter } from "./frontmatter";

describe("parseFrontmatter", () => {
  it("parses single-line values", () => {
    const { frontmatter, body } = parseFrontmatter(
      "---\nname: foo\ndescription: bar\n---\nbody content",
    );
    expect(frontmatter).toEqual({ name: "foo", description: "bar" });
    expect(body).toBe("body content");
  });

  it("strips double quotes", () => {
    const { frontmatter } = parseFrontmatter(
      '---\nname: "foo"\ndescription: "hello world"\n---\n',
    );
    expect(frontmatter).toEqual({ name: "foo", description: "hello world" });
  });

  it("strips single quotes", () => {
    const { frontmatter } = parseFrontmatter(
      "---\nname: 'foo'\ndescription: 'hello world'\n---\n",
    );
    expect(frontmatter).toEqual({ name: "foo", description: "hello world" });
  });

  it("preserves newlines in literal block scalar", () => {
    const { frontmatter } = parseFrontmatter(
      "---\nname: foo\ndescription: |\n  line1\n  line2\n---\n",
    );
    expect(frontmatter?.description).toBe("line1\nline2\n");
  });

  it("drops trailing newline with strip chomping", () => {
    const { frontmatter } = parseFrontmatter(
      "---\nname: foo\ndescription: |-\n  line1\n  line2\n---\n",
    );
    expect(frontmatter?.description).toBe("line1\nline2");
  });

  it("joins folded block scalar with spaces", () => {
    const { frontmatter } = parseFrontmatter(
      "---\nname: foo\ndescription: >\n  line1\n  line2\n---\n",
    );
    expect(frontmatter?.description).toBe("line1 line2\n");
  });

  it("handles CRLF line endings", () => {
    const { frontmatter } = parseFrontmatter(
      "---\r\nname: foo\r\ndescription: bar\r\n---\r\nbody",
    );
    expect(frontmatter).toEqual({ name: "foo", description: "bar" });
  });

  it("returns null when no frontmatter", () => {
    const result = parseFrontmatter("plain markdown");
    expect(result.frontmatter).toBeNull();
    expect(result.body).toBe("plain markdown");
  });

  it("returns null when frontmatter is unterminated", () => {
    const result = parseFrontmatter("---\nname: foo\ndescription: bar\n");
    expect(result.frontmatter).toBeNull();
  });

  it("returns null when YAML is invalid", () => {
    const result = parseFrontmatter("---\n: : : not valid\n---\nbody");
    expect(result.frontmatter).toBeNull();
  });

  it("reproduces issue 3495 with Chinese block scalar", () => {
    const content =
      "---\n" +
      "name: requirements-workshop\n" +
      "description: |\n" +
      "  当用户想要开发新功能、讨论需求、梳理业务逻辑时触发。通过多轮对话将模糊的想法转化为结构化的需求文档，为后续技术方案设计提供输入。\n" +
      '  适用场景：用户说"我想实现XX功能"、"讨论一下这个需求"、"帮我分析一下怎么做"、"这个需求该怎么设计"、"写个需求文档"等。\n' +
      "  本skill只负责产出需求文档，不进入技术设计或代码实现阶段。\n" +
      "---\nbody";

    const { frontmatter } = parseFrontmatter(content);
    expect(frontmatter?.name).toBe("requirements-workshop");
    expect(frontmatter?.description).toBe(
      "当用户想要开发新功能、讨论需求、梳理业务逻辑时触发。通过多轮对话将模糊的想法转化为结构化的需求文档，为后续技术方案设计提供输入。\n" +
        '适用场景：用户说"我想实现XX功能"、"讨论一下这个需求"、"帮我分析一下怎么做"、"这个需求该怎么设计"、"写个需求文档"等。\n' +
        "本skill只负责产出需求文档，不进入技术设计或代码实现阶段。\n",
    );
  });

  it("coerces non-string values to strings", () => {
    const { frontmatter } = parseFrontmatter(
      "---\nname: foo\nversion: 1\nenabled: true\n---\n",
    );
    expect(frontmatter).toEqual({
      name: "foo",
      version: "1",
      enabled: "true",
    });
  });

  // A structured value where a scalar belongs (authoring mistake) must keep the
  // sibling name and JSON-encode the value. This mirrors the Go
  // ParseSkillFrontmatter behaviour in server/internal/skill so both sides agree.
  it("keeps sibling name and JSON-encodes a sequence value", () => {
    const { frontmatter } = parseFrontmatter(
      "---\nname: my-skill\ndescription:\n  - first feature\n  - second feature\n---\nbody",
    );
    expect(frontmatter).toEqual({
      name: "my-skill",
      description: '["first feature","second feature"]',
    });
  });

  it("keeps sibling name and JSON-encodes a mapping value", () => {
    const { frontmatter } = parseFrontmatter(
      "---\nname: my-skill\ndescription:\n  a: 1\n  b: 2\n---\nbody",
    );
    expect(frontmatter).toEqual({
      name: "my-skill",
      description: '{"a":1,"b":2}',
    });
  });
});

File: server/internal/skill/frontmatter_test.go
package skill

import "testing"

func TestParseSkillFrontmatter(t *testing.T) {
	tests := []struct {
		name     string
		content  string
		wantName string
		wantDesc string
	}{
		{
			name:     "single line",
			content:  "---\nname: foo\ndescription: bar\n---\nbody",
			wantName: "foo",
			wantDesc: "bar",
		},
		{
			name:     "double quoted",
			content:  "---\nname: \"foo\"\ndescription: \"hello world\"\n---\nbody",
			wantName: "foo",
			wantDesc: "hello world",
		},
		{
			name:     "single quoted",
			content:  "---\nname: 'foo'\ndescription: 'hello world'\n---\nbody",
			wantName: "foo",
			wantDesc: "hello world",
		},
		{
			name:     "literal block scalar keeps newlines",
			content:  "---\nname: foo\ndescription: |\n  line1\n  line2\n---\nbody",
			wantName: "foo",
			wantDesc: "line1\nline2\n",
		},
		{
			name:     "literal strip chomping drops trailing newline",
			content:  "---\nname: foo\ndescription: |-\n  line1\n  line2\n---\nbody",
			wantName: "foo",
			wantDesc: "line1\nline2",
		},
		{
			name:     "folded block scalar joins with spaces",
			content:  "---\nname: foo\ndescription: >\n  line1\n  line2\n---\nbody",
			wantName: "foo",
			wantDesc: "line1 line2\n",
		},
		{
			name:     "CRLF line endings",
			content:  "---\r\nname: foo\r\ndescription: bar\r\n---\r\nbody",
			wantName: "foo",
			wantDesc: "bar",
		},
		{
			name:     "no frontmatter returns empty",
			content:  "no frontmatter here",
			wantName: "",
			wantDesc: "",
		},
		{
			name:     "unterminated frontmatter returns empty",
			content:  "---\nname: foo\ndescription: bar\n",
			wantName: "",
			wantDesc: "",
		},
		{
			name:     "invalid YAML falls back to empty",
			content:  "---\n: : : not valid\n---\nbody",
			wantName: "",
			wantDesc: "",
		},
		{
			name:     "name only",
			content:  "---\nname: foo\n---\nbody",
			wantName: "foo",
			wantDesc: "",
		},
		{
			name:     "description only",
			content:  "---\ndescription: bar\n---\nbody",
			wantName: "",
			wantDesc: "bar",
		},
		{
			name:     "leading blank line is not frontmatter",
			content:  "\n---\nname: foo\ndescription: bar\n---\nbody",
			wantName: "",
			wantDesc: "",
		},
		{
			// The non-greedy capture stops at the first closing fence, so a
			// later "---" in the body must not extend the frontmatter block.
			name:     "triple dash in body stops at first fence",
			content:  "---\nname: foo\ndescription: bar\n---\nintro\n---\nmore",
			wantName: "foo",
			wantDesc: "bar",
		},
		{
			// Parity with the TS coercion: non-string scalars render as their
			// literal form rather than being dropped.
			name:     "non-string scalars coerce to literal",
			content:  "---\nname: 123\ndescription: 456\n---\nbody",
			wantName: "123",
			wantDesc: "456",
		},
		{
			// A structured value where a scalar belongs (authoring mistake) must
			// not discard the sibling name; the value is JSON-encoded, matching
			// the TS parseFrontmatter behaviour.
			name:     "sequence description keeps name and is JSON-encoded",
			content:  "---\nname: my-skill\ndescription:\n  - first feature\n  - second feature\n---\nbody",
			wantName: "my-skill",
			wantDesc: `["first feature","second feature"]`,
		},
		{
			name:     "mapping description keeps name and is JSON-encoded",
			content:  "---\nname: my-skill\ndescription:\n  a: 1\n  b: 2\n---\nbody",
			wantName: "my-skill",
			wantDesc: `{"a":1,"b":2}`,
		},
		{
			// Reproduction for issue #3495: Chinese block scalar.
			name: "issue 3495 chinese literal block scalar",
			content: "---\n" +
				"name: requirements-workshop\n" +
				"description: |\n" +
				"  当用户想要开发新功能、讨论需求、梳理业务逻辑时触发。通过多轮对话将模糊的想法转化为结构化的需求文档，为后续技术方案设计提供输入。\n" +
				"  适用场景：用户说\"我想实现XX功能\"、\"讨论一下这个需求\"、\"帮我分析一下怎么做\"、\"这个需求该怎么设计\"、\"写个需求文档\"等。\n" +
				"  本skill只负责产出需求文档，不进入技术设计或代码实现阶段。\n" +
				"---\nbody",
			wantName: "requirements-workshop",
			wantDesc: "当用户想要开发新功能、讨论需求、梳理业务逻辑时触发。通过多轮对话将模糊的想法转化为结构化的需求文档，为后续技术方案设计提供输入。\n" +
				"适用场景：用户说\"我想实现XX功能\"、\"讨论一下这个需求\"、\"帮我分析一下怎么做\"、\"这个需求该怎么设计\"、\"写个需求文档\"等。\n" +
				"本skill只负责产出需求文档，不进入技术设计或代码实现阶段。\n",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			gotName, gotDesc := ParseSkillFrontmatter(tt.content)
			if gotName != tt.wantName {
				t.Errorf("name: got %q, want %q", gotName, tt.wantName)
			}
			if gotDesc != tt.wantDesc {
				t.Errorf("description: got %q, want %q", gotDesc, tt.wantDesc)
			}
		})
	}
}
